photon-ml/src/main/scala/com/linkedin/photon/ml/ModelTraining.scala

```diff
@@ -155,42 +163,36 @@ object ModelTraining extends Logging {
       s"${warmStartModels.keys.mkString(", ")}")
 
     val initWeightsAndModels = List[(Double, GeneralizedLinearModel)]()
-    val (finalOptimizationProblem, finalWeightsAndModels) = sortedRegularizationWeights
-      .foldLeft((initOptimizationProblem, initWeightsAndModels)) {
-        case ((optimizationProblem, List()), currentWeight) =>
+    val finalWeightsAndModels = sortedRegularizationWeights
+      .foldLeft(initWeightsAndModels) {
+        case (List(), currentWeight) =>
           // Initialize the list with the result from the first regularization weight
-          val updatedOptimizationProblem = optimizationProblem.updateObjective(
-            normalizationContextProvider, currentWeight)
+          optimizationProblem.updateRegularizationWeight(currentWeight)
 
           if (numWarmStartModels == 0) {
             logInfo(s"No warm start model found; beginning training with a 0-coefficients model")
 
-            (updatedOptimizationProblem,
-              List((currentWeight, updatedOptimizationProblem.run(trainingData, normalizationContext))))
+              List((currentWeight, optimizationProblem.run(trainingData)))
           } else {
             val maxLambda = warmStartModels.keys.max
             logInfo(s"Starting training using warm-start model with lambda = $maxLambda")
 
-            (updatedOptimizationProblem,
-              List((currentWeight,
-                updatedOptimizationProblem.run(
-                  trainingData, warmStartModels.get(maxLambda).get, normalizationContext))))
+            List((currentWeight, optimizationProblem.run(trainingData, warmStartModels.get(maxLambda).get)))
           }
 
-        case ((latestOptimizationProblem, latestWeightsAndModels), currentWeight) =>
+        case (latestWeightsAndModels, currentWeight) =>
           // Train the rest of the models
           val previousModel = latestWeightsAndModels.head._2
-          val updatedOptimizationProblem = latestOptimizationProblem.updateObjective(
-            normalizationContextProvider, currentWeight)
+          optimizationProblem.updateRegularizationWeight(currentWeight)
 
           logInfo(s"Training model with regularization weight $currentWeight finished")
 
-          (updatedOptimizationProblem,
-            (currentWeight, updatedOptimizationProblem.run(trainingData, previousModel, normalizationContext))
-              :: latestWeightsAndModels)
+          (currentWeight, optimizationProblem.run(trainingData, previousModel)) +: latestWeightsAndModels
```

Comment:
+@TDDFT

@ashelkovnykov Can you briefly describe the changes made with feature normalization context?

Since we have plans to support feature normalization/scaling in GAME soon, it would be great if you can help bring folks to the same page.